A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,232 @@
+  if (header == null) {
+    using_random_hash = true;
+    request_hash = ComputeRandomHash();
+  } else {
+    request_hash = ComputeHash(header);
+  }
+}
+
+first_index = ring.FindIndexForHash(request_hash);
+if !using_random_hash {
+    // Use the logic from A62 unchanged.
+    // ...
+}
+
+requested_connection = picker_has_a_child_connecting;
+for (i = 0; i < ring.size(); ++i) {
+  index = (first_index + i) % ring.size();
+  if (ring[index].state == READY) {
+    return ring[index].picker->Pick(...);
+  }
+  if (!requested_connection && ring[index].state == IDLE) {
+    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
+    requested_connection = true;
+  }
+}
+if (requested_connection) return PICK_QUEUE;
+return PICK_FAIL;
+```
+
+This behavior ensures that a single RPC does not cause more than one endpoint to
+exit `IDLE` state at a time, and that a request missing the header does not
+incur extra latency in the common case where there is already at least one
+endpoint in `READY` state. It converges to picking a random endpoint, since each
+request may eventually cause a random endpoint to go from `IDLE` to `READY`.
+
+### Explicitly setting the endpoint hash key
+
+The `ring_hash` policy will be changed such that the hash key used for
+determining the locations of each endpoint on the ring will be extracted from a
+pre-defined endpoint attribute called `hash_key`. If this attribute is set, then
+the endpoint is placed on the ring by hashing its value. If this attribute is
+not set or empty, then the endpoint's first address is hashed, matching the
+current behavior. The locations of an existing endpoint on the ring is updated
+if its `hash_key` endpoint attribute changes.
+
+The xDS resolver, described in [A74][A74], will be changed to set the `hash_key`
+endpoint attribute to the value of [LbEndpoint.Metadata][LbEndpoint.Metadata]
+`envoy.lb` `hash_key` field, as described in [Envoy's documentation for the ring
+hash load balancer][envoy-ringhash].
+
+### Temporary environment variable protection
+
+Explicitly setting the request hash key will be gated by the
+`GRPC_EXPERIMENTAL_RING_HASH_SET_REQUEST_HASH_KEY` environment variable. This
+mechanism will be supported for a couple of gRPC releases but will be removed in
+the long run.
+
+Adding support for the `hash_key` in xDS endpoint metadata could potentially break
+existing clients whose control plane is setting this key, because upgrading the
+client to a new version of gRPC would automatically cause the key to start being
+used. We expect that this change will not cause problems for most users, but
+just in case there is a problem, we will provide a migration story by supporting
+a temporary mechanism to tell gRPC to ignore the `hash_key` endpoint
+metadata. This will be enabled by setting the
+`GRPC_XDS_ENDPOINT_HASH_KEY_BACKWARD_COMPAT` environment variable to true. This
+mechanism will be supported for a couple of gRPC releases but will be removed in
+the long run.
```

Comment:
Should we have a rollout plan here?  I.e. something like "The first release will have it off by default.  A subsequent release will enable it by default, but leave the opt-out.  A final release will remove the opt-out."

Reply:
I'm fine either way, so I took your suggestion. The wording here was based on Mark's suggestion in https://github.com/grpc/proposal/pull/412#discussion_r1462478950.